# MLC Project - Napster
Group Focus: Data provided from DSPS

Considerations: which DSPs are doing a bad job of sending over complete or at least minimum data required for matching?

Which DSPs have the most streams? Do songs get more air time on certain DSPs?

Does song length correlate to DSP in any way?

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import re

## Read data into notebook

In [2]:
mlc = pd.read_csv('../data/MLC_sample.csv')
mlc.head()

If  Country Code Registrant Code  Year of Reference Usage Period  \
0  USUG12101043           US             UG1               21.0          NaN   
1  USRC12100543           US             RC1               21.0          NaN   
2  USSM12102263           US             SM1               21.0          NaN   
3  USLD91731547           US             LD9               17.0     2/1/2021   
4  USAT22007048           US             AT2               20.0          NaN   

  Streaming Platform (DSP)  Streams        Recording Artist Recording Label  \
0                AudioMack    10175                  Future             NaN   
1               SoundCloud     8597                LUCKY3RD        LUCKY3RD   
2               SoundCloud   261280                LUCKY3RD        LUCKY3RD   
3                   Trebel        5  Bachata & Merengue Mix         Orchard   
4                AudioMack    62105                  Foolio             NaN   

                             Recording Title            Release Title  \
0   FUTURE FT DEJ LOAF HEY THERE PROD BY DDS                #unknown#   
1                     Keep It Cool  LUCKY3RD   Keep It Cool  LUCKY3RD   
2                     Life Goes On  LUCKY3RD   Life Goes On  LUCKY3RD   
3     No dudes de mi- Merengue & Bachata Mix            Mega Mix 2010   
4                       WHEN I SEE YOU REMIX                #unknown#   

   Recording Duration (Seconds)  Songwriter's Listed (1 = "Yes", 0 = "No")  
0                           181                                          0  
1                           133                                          0  
2                           171                                          0  
3                          1250                                          0  
4                           187                                          0

## Rename columns

In [3]:
mlc.columns = ['ISRC', 'Country_Code', 'Registrant_Code', 'Year_of_Reference', 'Usage_Period', 'DSP', 'Streams', 'Recording_Artist', 'Recording_Label', 'Recording_Title', 'Release_Title', 'Recording_Duration_(Seconds)', 'Songwriters_Listed_(1=Y_0=N)']
mlc.head()

ISRC Country_Code Registrant_Code  Year_of_Reference Usage_Period  \
0  USUG12101043           US             UG1               21.0          NaN   
1  USRC12100543           US             RC1               21.0          NaN   
2  USSM12102263           US             SM1               21.0          NaN   
3  USLD91731547           US             LD9               17.0     2/1/2021   
4  USAT22007048           US             AT2               20.0          NaN   

          DSP  Streams        Recording_Artist Recording_Label  \
0   AudioMack    10175                  Future             NaN   
1  SoundCloud     8597                LUCKY3RD        LUCKY3RD   
2  SoundCloud   261280                LUCKY3RD        LUCKY3RD   
3      Trebel        5  Bachata & Merengue Mix         Orchard   
4   AudioMack    62105                  Foolio             NaN   

                             Recording_Title            Release_Title  \
0   FUTURE FT DEJ LOAF HEY THERE PROD BY DDS                #unknown#   
1                     Keep It Cool  LUCKY3RD   Keep It Cool  LUCKY3RD   
2                     Life Goes On  LUCKY3RD   Life Goes On  LUCKY3RD   
3     No dudes de mi- Merengue & Bachata Mix            Mega Mix 2010   
4                       WHEN I SEE YOU REMIX                #unknown#   

   Recording_Duration_(Seconds)  Songwriters_Listed_(1=Y_0=N)  
0                           181                             0  
1                           133                             0  
2                           171                             0  
3                          1250                             0  
4                           187                             0

## Create column(s) for song length details

In [4]:
def assign_length(row):
    if row <= 120:
        result = "Up to 2 min"
    elif row <= 240:
        result = "2-4 min"
    elif row <= 420:
        result = "4-7 min"
    elif row <= 600:
        result = "7-10 min"
    elif row <= 900:
        result = "10-15 min"
    elif row <= 3600:
        result = "15-60 min"
    elif row <= 86400:
        result = "1-24 hours"
    else:
        result = "More than 1 day"
    return result

mlc['Song_Length'] = mlc['Recording_Duration_(Seconds)'].apply(assign_length)
mlc.head()

ISRC Country_Code Registrant_Code  Year_of_Reference Usage_Period  \
0  USUG12101043           US             UG1               21.0          NaN   
1  USRC12100543           US             RC1               21.0          NaN   
2  USSM12102263           US             SM1               21.0          NaN   
3  USLD91731547           US             LD9               17.0     2/1/2021   
4  USAT22007048           US             AT2               20.0          NaN   

          DSP  Streams        Recording_Artist Recording_Label  \
0   AudioMack    10175                  Future             NaN   
1  SoundCloud     8597                LUCKY3RD        LUCKY3RD   
2  SoundCloud   261280                LUCKY3RD        LUCKY3RD   
3      Trebel        5  Bachata & Merengue Mix         Orchard   
4   AudioMack    62105                  Foolio             NaN   

                             Recording_Title            Release_Title  \
0   FUTURE FT DEJ LOAF HEY THERE PROD BY DDS                #unknown#   
1                     Keep It Cool  LUCKY3RD   Keep It Cool  LUCKY3RD   
2                     Life Goes On  LUCKY3RD   Life Goes On  LUCKY3RD   
3     No dudes de mi- Merengue & Bachata Mix            Mega Mix 2010   
4                       WHEN I SEE YOU REMIX                #unknown#   

   Recording_Duration_(Seconds)  Songwriters_Listed_(1=Y_0=N) Song_Length  
0                           181                             0     2-4 min  
1                           133                             0     2-4 min  
2                           171                             0     2-4 min  
3                          1250                             0   15-60 min  
4                           187                             0     2-4 min

In [5]:
def compared_to_median(row):
    if row < 190:
        result = "Lower"
    elif row == 190:
        result = "Median"
    else:
        result = "Higher"
    return result
mlc['Length_Compared_to_Median'] = mlc['Recording_Duration_(Seconds)'].apply(compared_to_median)
mlc.head()

ISRC Country_Code Registrant_Code  Year_of_Reference Usage_Period  \
0  USUG12101043           US             UG1               21.0          NaN   
1  USRC12100543           US             RC1               21.0          NaN   
2  USSM12102263           US             SM1               21.0          NaN   
3  USLD91731547           US             LD9               17.0     2/1/2021   
4  USAT22007048           US             AT2               20.0          NaN   

          DSP  Streams        Recording_Artist Recording_Label  \
0   AudioMack    10175                  Future             NaN   
1  SoundCloud     8597                LUCKY3RD        LUCKY3RD   
2  SoundCloud   261280                LUCKY3RD        LUCKY3RD   
3      Trebel        5  Bachata & Merengue Mix         Orchard   
4   AudioMack    62105                  Foolio             NaN   

                             Recording_Title            Release_Title  \
0   FUTURE FT DEJ LOAF HEY THERE PROD BY DDS                #unknown#   
1                     Keep It Cool  LUCKY3RD   Keep It Cool  LUCKY3RD   
2                     Life Goes On  LUCKY3RD   Life Goes On  LUCKY3RD   
3     No dudes de mi- Merengue & Bachata Mix            Mega Mix 2010   
4                       WHEN I SEE YOU REMIX                #unknown#   

   Recording_Duration_(Seconds)  Songwriters_Listed_(1=Y_0=N) Song_Length  \
0                           181                             0     2-4 min   
1                           133                             0     2-4 min   
2                           171                             0     2-4 min   
3                          1250                             0   15-60 min   
4                           187                             0     2-4 min   

  Length_Compared_to_Median  
0                     Lower  
1                     Lower  
2                     Lower  
3                    Higher  
4                     Lower

## Check for null values in each column

In [6]:
mlc.isnull().sum()

ISRC                            1760
Country_Code                    1697
Registrant_Code                 1697
Year_of_Reference               1761
Usage_Period                    8102
DSP                             3999
Streams                            0
Recording_Artist                   0
Recording_Label                 1008
Recording_Title                    0
Release_Title                     69
Recording_Duration_(Seconds)       0
Songwriters_Listed_(1=Y_0=N)       0
Song_Length                        0
Length_Compared_to_Median          0
dtype: int64

## Check data type in each column

In [7]:
mlc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 15 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   ISRC                          98240 non-null   object 
 1   Country_Code                  98303 non-null   object 
 2   Registrant_Code               98303 non-null   object 
 3   Year_of_Reference             98239 non-null   float64
 4   Usage_Period                  91898 non-null   object 
 5   DSP                           96001 non-null   object 
 6   Streams                       100000 non-null  int64  
 7   Recording_Artist              100000 non-null  object 
 8   Recording_Label               98992 non-null   object 
 9   Recording_Title               100000 non-null  object 
 10  Release_Title                 99931 non-null   object 
 11  Recording_Duration_(Seconds)  100000 non-null  int64  
 12  Songwriters_Listed_(1=Y_0=N)  100000 non-null

## Count of DSPs

In [8]:
mlc['DSP'].value_counts()

Spotify               32268
Apple                 22200
Amazon                14438
Pandora               13777
Tidal                  3521
YouTube                2752
SoundCloud             2122
GTL                    1090
Melodyv                 829
Trebel                  817
iHeart Radio            707
AudioMack               550
NugsNet                 316
LiveXLive               200
Qoboz                   104
Midwest Tape            102
Deezer                   51
Anghami                  46
Sonos                    23
Recisio                  22
Smithsonian              21
Ultimate Guitar          19
PowerMusic                8
Wolfgangs                 4
Fan Label                 4
MixCloud                  4
Pacemaker                 3
Classical Archives        2
MonkingMe                 1
Name: DSP, dtype: int64

In [9]:
DSP_list = mlc['DSP'].unique()
print(DSP_list)

['AudioMack' 'SoundCloud' 'Trebel' 'Spotify' 'Apple' 'Amazon' 'Pandora'
 'YouTube' nan 'Tidal' 'iHeart Radio' 'Melodyv' 'Deezer' 'GTL' 'LiveXLive'
 'NugsNet' 'Midwest Tape' 'Ultimate Guitar' 'Qoboz' 'Anghami' 'Pacemaker'
 'PowerMusic' 'Wolfgangs' 'Recisio' 'Sonos' 'Smithsonian' 'Fan Label'
 'MixCloud' 'MonkingMe' 'Classical Archives']


def main():

for x in DSP_list (or mlc['DSP'])?

Need to either reset the index on the df OR use orginal df based on the DSP column and then run a unique to avoid repeats. That will create the list that can then be used for the for loop.

There are 29 DSPs with at least one stream attributed, but there are 3999 entries in the dataset that have a null value for DSP.

## DSP DataFrames for further analysis

In [10]:
deezer = mlc.loc[mlc['DSP'] == 'Deezer']
deezer.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 5
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [11]:
anghami = mlc.loc[mlc['DSP'] == 'Anghami']
anghami.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 3
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [12]:
sonos = mlc.loc[mlc['DSP'] == 'Sonos']
sonos.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [13]:
recisio = mlc.loc[mlc['DSP'] == 'Recisio']
recisio.isnull().sum()

ISRC                             0
Country_Code                     0
Registrant_Code                  0
Year_of_Reference                0
Usage_Period                     0
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                 22
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [14]:
smithsonian = mlc.loc[mlc['DSP'] == 'Smithsonian']
smithsonian.isnull().sum()

ISRC                            1
Country_Code                    1
Registrant_Code                 1
Year_of_Reference               1
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [15]:
ultimate_guitar = mlc.loc[mlc['DSP'] == 'Ultimate Guitar']
ultimate_guitar.isnull().sum()

ISRC                             0
Country_Code                     0
Registrant_Code                  0
Year_of_Reference                0
Usage_Period                     0
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                 19
Recording_Title                  0
Release_Title                   19
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [16]:
powermusic = mlc.loc[mlc['DSP'] == 'PowerMusic']
powermusic.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 8
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [17]:
wolfgangs = mlc.loc[mlc['DSP'] == 'Wolfgangs']
wolfgangs.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 4
Recording_Title                 0
Release_Title                   4
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [18]:
fan_label = mlc.loc[mlc['DSP'] == 'Fan Label']
fan_label.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [19]:
mixcloud = mlc.loc[mlc['DSP'] == 'MixCloud']
mixcloud.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 4
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [20]:
pacemaker = mlc.loc[mlc['DSP'] == 'Pacemaker']
pacemaker.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [21]:
classical_archives = mlc.loc[mlc['DSP'] == 'Classical Archives']
classical_archives.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [22]:
monkingme = mlc.loc[mlc['DSP'] =='MonkingMe']
monkingme.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 1
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [23]:
unknownDSP = mlc.loc[mlc['DSP'].isna()]
unknownDSP.isnull().sum()

ISRC                              58
Country_Code                      54
Registrant_Code                   54
Year_of_Reference                 58
Usage_Period                    3999
DSP                             3999
Streams                            0
Recording_Artist                   0
Recording_Label                   47
Recording_Title                    0
Release_Title                     35
Recording_Duration_(Seconds)       0
Songwriters_Listed_(1=Y_0=N)       0
Song_Length                        0
Length_Compared_to_Median          0
dtype: int64

In [24]:
spotify = mlc.loc[mlc['DSP'] == 'Spotify']
spotify.isnull().sum()

ISRC                            571
Country_Code                    554
Registrant_Code                 554
Year_of_Reference               571
Usage_Period                      0
DSP                               0
Streams                           0
Recording_Artist                  0
Recording_Label                  11
Recording_Title                   0
Release_Title                     1
Recording_Duration_(Seconds)      0
Songwriters_Listed_(1=Y_0=N)      0
Song_Length                       0
Length_Compared_to_Median         0
dtype: int64

In [25]:
apple = mlc.loc[mlc['DSP'] == 'Apple']
apple.isnull().sum()

ISRC                             380
Country_Code                     363
Registrant_Code                  363
Year_of_Reference                380
Usage_Period                    1441
DSP                                0
Streams                            0
Recording_Artist                   0
Recording_Label                    8
Recording_Title                    0
Release_Title                      7
Recording_Duration_(Seconds)       0
Songwriters_Listed_(1=Y_0=N)       0
Song_Length                        0
Length_Compared_to_Median          0
dtype: int64

In [26]:
pandora = mlc.loc[mlc['DSP'] == 'Pandora']
pandora.isnull().sum()

ISRC                            237
Country_Code                    230
Registrant_Code                 230
Year_of_Reference               238
Usage_Period                      0
DSP                               0
Streams                           0
Recording_Artist                  0
Recording_Label                  27
Recording_Title                   0
Release_Title                     3
Recording_Duration_(Seconds)      0
Songwriters_Listed_(1=Y_0=N)      0
Song_Length                       0
Length_Compared_to_Median         0
dtype: int64

In [27]:
tidal = mlc.loc[mlc['DSP'] == 'Tidal']
tidal.isnull().sum()

ISRC                             53
Country_Code                     52
Registrant_Code                  52
Year_of_Reference                53
Usage_Period                    868
DSP                               0
Streams                           0
Recording_Artist                  0
Recording_Label                   0
Recording_Title                   0
Release_Title                     0
Recording_Duration_(Seconds)      0
Songwriters_Listed_(1=Y_0=N)      0
Song_Length                       0
Length_Compared_to_Median         0
dtype: int64

In [28]:
youtube = mlc.loc[mlc['DSP'] == 'YouTube']
youtube.isnull().sum()

ISRC                             56
Country_Code                     54
Registrant_Code                  54
Year_of_Reference                56
Usage_Period                    958
DSP                               0
Streams                           0
Recording_Artist                  0
Recording_Label                 233
Recording_Title                   0
Release_Title                     0
Recording_Duration_(Seconds)      0
Songwriters_Listed_(1=Y_0=N)      0
Song_Length                       0
Length_Compared_to_Median         0
dtype: int64

In [29]:
soundcloud = mlc.loc[mlc['DSP'] == 'SoundCloud']
soundcloud.isnull().sum()

ISRC                             36
Country_Code                     36
Registrant_Code                  36
Year_of_Reference                36
Usage_Period                    494
DSP                               0
Streams                           0
Recording_Artist                  0
Recording_Label                   0
Recording_Title                   0
Release_Title                     0
Recording_Duration_(Seconds)      0
Songwriters_Listed_(1=Y_0=N)      0
Song_Length                       0
Length_Compared_to_Median         0
dtype: int64

In [30]:
gtl = mlc.loc[mlc['DSP'] == 'GTL']
gtl.isnull().sum()

ISRC                            26
Country_Code                    24
Registrant_Code                 24
Year_of_Reference               26
Usage_Period                     0
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                  0
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [31]:
melodyv = mlc.loc[mlc['DSP'] == 'Melodyv']
melodyv.isnull().sum()

ISRC                            15
Country_Code                    15
Registrant_Code                 15
Year_of_Reference               15
Usage_Period                     0
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                  0
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [32]:
trebel = mlc.loc[mlc['DSP'] == 'Trebel']
trebel.isnull().sum()

ISRC                            19
Country_Code                    18
Registrant_Code                 18
Year_of_Reference               19
Usage_Period                     0
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                  0
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [33]:
iheartradio = mlc.loc[mlc['DSP'] == 'iHeart Radio']
iheartradio.isnull().sum()

ISRC                            16
Country_Code                    16
Registrant_Code                 16
Year_of_Reference               16
Usage_Period                    38
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                  0
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [34]:
audiomack = mlc.loc[mlc['DSP'] == 'AudioMack']
audiomack.isnull().sum()

ISRC                             16
Country_Code                     15
Registrant_Code                  15
Year_of_Reference                16
Usage_Period                    287
DSP                               0
Streams                           0
Recording_Artist                  0
Recording_Label                 550
Recording_Title                   0
Release_Title                     0
Recording_Duration_(Seconds)      0
Songwriters_Listed_(1=Y_0=N)      0
Song_Length                       0
Length_Compared_to_Median         0
dtype: int64

In [35]:
nugsnet = mlc.loc[mlc['DSP'] == 'NugsNet']
nugsnet.isnull().sum()

ISRC                            4
Country_Code                    3
Registrant_Code                 3
Year_of_Reference               4
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [36]:
livexlive = mlc.loc[mlc['DSP'] == 'LiveXLive']
livexlive.isnull().sum()

ISRC                             4
Country_Code                     4
Registrant_Code                  4
Year_of_Reference                4
Usage_Period                    11
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                 24
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [37]:
qoboz = mlc.loc[mlc['DSP'] == 'Qoboz']
qoboz.isnull().sum()

ISRC                            0
Country_Code                    0
Registrant_Code                 0
Year_of_Reference               0
Usage_Period                    0
DSP                             0
Streams                         0
Recording_Artist                0
Recording_Label                 0
Recording_Title                 0
Release_Title                   0
Recording_Duration_(Seconds)    0
Songwriters_Listed_(1=Y_0=N)    0
Song_Length                     0
Length_Compared_to_Median       0
dtype: int64

In [38]:
midwesttape = mlc.loc[mlc['DSP'] == 'Midwest Tape']
midwesttape.isnull().sum()

ISRC                             3
Country_Code                     3
Registrant_Code                  3
Year_of_Reference                3
Usage_Period                     6
DSP                              0
Streams                          0
Recording_Artist                 0
Recording_Label                 42
Recording_Title                  0
Release_Title                    0
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0
dtype: int64

In [39]:
pd.DataFrame(unknownDSP.isnull().sum())

0
ISRC                            58
Country_Code                    54
Registrant_Code                 54
Year_of_Reference               58
Usage_Period                  3999
DSP                           3999
Streams                          0
Recording_Artist                 0
Recording_Label                 47
Recording_Title                  0
Release_Title                   35
Recording_Duration_(Seconds)     0
Songwriters_Listed_(1=Y_0=N)     0
Song_Length                      0
Length_Compared_to_Median        0

## Deeper dive into song length details
Dataset provides song duration in seconds

In [40]:
mlc['Recording_Duration_(Seconds)'].describe()

count    100000.000000
mean       1016.893690
std       15565.692133
min           0.000000
25%         149.000000
50%         190.000000
75%         236.000000
max      818738.000000
Name: Recording_Duration_(Seconds), dtype: float64

In [41]:
mlc.nlargest(10, ['Recording_Duration_(Seconds)'])

ISRC Country_Code Registrant_Code  Year_of_Reference  \
19931  QZES81895037           QZ             ES8               18.0   
34584  DEN962000842           DE             N96               20.0   
20399  JMA272011862           JM             A27               20.0   
48381  USAT22006112           US             AT2               20.0   
138    USUYG1360358           US             UYG               13.0   
139    USUM72105617           US             UM7               21.0   
140    USAT22007802           US             AT2               20.0   
141    USUM72024668           US             UM7               20.0   
142    USUG12004700           US             UG1               20.0   
258    QZMGB2100021           QZ             MGB               21.0   

      Usage_Period      DSP  Streams     Recording_Artist Recording_Label  \
19931     2/1/2021   Trebel     6464  Megan Thee Stallion          Warner   
34584     2/1/2021   Trebel     4797                  SZA      Sony Music   
20399     2/1/2021   Trebel     3492               H.E.R.      Sony Music   
48381     2/1/2021   Trebel     2965           Kevo Muney          Warner   
138       4/1/2021   Deezer        2              Unknown             NaN   
139       4/1/2021   Deezer        2              Unknown             NaN   
140       5/1/2021   Deezer        2              Unknown             NaN   
141       4/1/2021   Deezer        2              Unknown             NaN   
142       4/1/2021   Deezer        2              Unknown             NaN   
258       3/1/2021  Pandora      160           Bizzy Bone      Bizzy Bone   

               Recording_Title                       Release_Title  \
19931  Cry Baby (feat. DaBaby)                           Good News   
34584                Good Days                           Good Days   
20399                   Damage                              Damage   
48381           Leave Some Day                      Leave Some Day   
138                {Cartagena}  Silvestre Dangond - Grandes Éxitos   
139                {Esa Mujer}  Silvestre Dangond - Grandes Éxitos   
140                {Esa Mujer}  Silvestre Dangond - Grandes Éxitos   
141       {Me Gusta, Me Gusta}  Silvestre Dangond - Grandes Éxitos   
142       {Me Gusta, Me Gusta}  Silvestre Dangond - Grandes Éxitos   
258                       0.44                          The Mantra   

       Recording_Duration_(Seconds)  Songwriters_Listed_(1=Y_0=N)  \
19931                        818738                             0   
34584                        770091                             0   
20399                        714437                             0   
48381                        463500                             0   
138                          356461                             0   
139                          356461                             0   
140                          356461                             0   
141                          356461                             0   
142                          356461                             0   
258                          356461                             1   

           Song_Length Length_Compared_to_Median  
19931  More than 1 day                    Higher  
34584  More than 1 day                    Higher  
20399  More than 1 day                    Higher  
48381  More than 1 day                    Higher  
138    More than 1 day                    Higher  
139    More than 1 day                    Higher  
140    More than 1 day                    Higher  
141    More than 1 day                    Higher  
142    More than 1 day                    Higher  
258    More than 1 day                    Higher

The longest song in this dataset comes in at 818,738 seconds which is equal to 227.4 hours. 356,461 also shows up in the dataset with more frequency than anticipated. This deserves some additional attention.

In [42]:
mlc.nsmallest(10, ['Recording_Duration_(Seconds)'])

ISRC Country_Code Registrant_Code  Year_of_Reference  \
14080  USAB02001303           US             AB0               20.0   
14081  GBKQU2111667           GB             KQU               21.0   
14916  AUHS01912771           AU             HS0               19.0   
33960  USRH12000094           US             RH1               20.0   
51916  QM24S1705172           QM             24S               17.0   
56892  USA6B0446310           US             A6B                4.0   
56893  USY251719876           US             Y25               17.0   
56894  USLR50100122           US             LR5                1.0   
81337  USWB12004709           US             WB1               20.0   
81634  QZMEN2057850           QZ             MEN               20.0   

      Usage_Period           DSP  Streams                  Recording_Artist  \
14080     3/1/2021         Apple    28962                  Mac-K the K Baby   
14081     3/1/2021         Apple    28962                  Mac-K the K Baby   
14916     2/1/2021  iHeart Radio     2192    Jelly Roll & Struggle Jennings   
33960     1/1/2021         Apple    43995                             T9ine   
51916     2/1/2021  iHeart Radio     1045    Jelly Roll & Struggle Jennings   
56892     3/1/2021         Apple   103842       RH Music, Eduardo Luzquiños   
56893     3/1/2021         Apple    20793       RH Music, Eduardo Luzquiños   
56894     3/1/2021         Apple    20793       RH Music, Eduardo Luzquiños   
81337     2/1/2021  iHeart Radio     1582                      Thom Rotella   
81634     3/1/2021         Apple    40593  Wovy, LowkeyLuke, Alejandro Lema   

            Recording_Label                    Recording_Title  \
14080        OneWay DumbWay                        Butterflies   
14081        OneWay DumbWay                        Butterflies   
14916  Struggle  Jelly Roll                      Can't Go Home   
33960              Columbia                          Go Harder   
51916  Struggle  Jelly Roll  Love Won (feat. Shooter Jennings)   
56892     Eduardo Luzquiños               Motive X Promiscuous   
56893     Eduardo Luzquiños               Motive X Promiscuous   
56894     Eduardo Luzquiños               Motive X Promiscuous   
81337          Thom Rotella                        Street Talk   
81634                  WOVY                  Stuntin' On My Ex   

                               Release_Title  Recording_Duration_(Seconds)  \
14080                   Butterflies - Single                             0   
14081                   Butterflies - Single                             0   
14916                      Waylon & Willie 2                             0   
33960                     Go Harder - Single                             0   
51916                      Waylon & Willie 2                             0   
56892  Motive X Promiscuous (Remix) - Single                             0   
56893  Motive X Promiscuous (Remix) - Single                             0   
56894  Motive X Promiscuous (Remix) - Single                             0   
81337                            Street Talk                             0   
81634             Stuntin' On My Ex - Single                             0   

       Songwriters_Listed_(1=Y_0=N)  Song_Length Length_Compared_to_Median  
14080                             1  Up to 2 min                     Lower  
14081                             1  Up to 2 min                     Lower  
14916                             0  Up to 2 min                     Lower  
33960                             1  Up to 2 min                     Lower  
51916                             0  Up to 2 min                     Lower  
56892                             1  Up to 2 min                     Lower  
56893                             1  Up to 2 min                     Lower  
56894                             1  Up to 2 min                     Lower  
81337                             0  Up to 2 min                     Lower  
81634  

There are a significant number of entries in the dataset with the recording duration listed as 0 seconds. Will seek additional clarity around these entries.

In [43]:
mlc['Recording_Duration_(Seconds)'].value_counts().loc[lambda x : x<60]

327      59
338      56
351      56
349      56
58       55
         ..
3794      1
33633     1
3894      1
15937     1
8755      1
Name: Recording_Duration_(Seconds), Length: 2000, dtype: int64

There are 2000 entries in the dataset with a song length less than 60 seconds.

In [44]:
# mlc['Recording Duration (Seconds)'].value_counts().loc[lambda x : x>60 and x<120]
# moving to a for loop with .apply in order to evaluate multiple conditions in a cleaner format

def assign_length(row):
    if row <= 120:
        result = "Up to 2 min"
    elif row <= 240:
        result = "2-4 min"
    elif row <= 420:
        result = "4-7 min"
    elif row <= 600:
        result = "7-10 min"
    elif row <= 900:
        result = "10-15 min"
    elif row <= 3600:
        result = "15-60 min"
    elif row <= 86400:
        result = "1-24 hours"
    else:
        result = "More than 1 day"
    return result

mlc['Song Length'] = mlc['Recording Duration (Seconds)'].apply(assign_length)
mlc.head()

In [45]:
mlc['Song_Length'].value_counts()

2-4 min            64179
4-7 min            17204
Up to 2 min        12487
7-10 min            2289
1-24 hours          1636
15-60 min           1257
10-15 min            751
More than 1 day      197
Name: Song_Length, dtype: int64

def compared_to_median(row):
    if row < 190:
        result = "Lower"
    elif row == 190:
        result = "Median"
    else:
        result = "Higher"
    return result
mlc['Length Compared to Median'] = mlc['Recording Duration (Seconds)'].apply(compared_to_median)
mlc.head()

In [46]:
mlc['Length_Compared_to_Median'].value_counts()

Lower     49891
Higher    49470
Median      639
Name: Length_Compared_to_Median, dtype: int64

In [47]:
mlc['Song_Length'].value_counts()

2-4 min            64179
4-7 min            17204
Up to 2 min        12487
7-10 min            2289
1-24 hours          1636
15-60 min           1257
10-15 min            751
More than 1 day      197
Name: Song_Length, dtype: int64

In [48]:
mlc['Song_Length'].describe()

count      100000
unique          8
top       2-4 min
freq        64179
Name: Song_Length, dtype: object

## Statistical insights

In [49]:
mlc.describe(include='all')

ISRC Country_Code Registrant_Code  Year_of_Reference  \
count          98240        98303           98303       98239.000000   
unique         54621           77            2368                NaN   
top     TCAEX2053645           US             UM7                NaN   
freq             119        41621            6020                NaN   
mean             NaN          NaN             NaN          19.671933   
std              NaN          NaN             NaN          12.638870   
min              NaN          NaN             NaN           0.000000   
25%              NaN          NaN             NaN          17.000000   
50%              NaN          NaN             NaN          19.000000   
75%              NaN          NaN             NaN          20.000000   
max              NaN          NaN             NaN          99.000000   

       Usage_Period      DSP       Streams Recording_Artist   Recording_Label  \
count         91898    96001  1.000000e+05           100000             98992   
unique            5       29           NaN            21149             11777   
top        3/1/2021  Spotify           NaN  Various artists  Atlantic Records   
freq          39887    32268           NaN              967              1384   
mean            NaN      NaN  5.626550e+04              NaN               NaN   
std             NaN      NaN  2.620276e+05              NaN               NaN   
min             NaN      NaN  1.000000e+00              NaN               NaN   
25%             NaN      NaN  6.807000e+03              NaN               NaN   
50%             NaN      NaN  1.677250e+04              NaN               NaN   
75%             NaN      NaN  3.990300e+04              NaN               NaN   
max             NaN      NaN  2.145617e+07              NaN               NaN   

       Recording_Title Release_Title  Recording_Duration_(Seconds)  \
count           100000         99931                 100000.000000   
unique           51057         36723                           NaN   
top             Savage     #unknown#                           NaN   
freq               134           231                           NaN   
mean               NaN           NaN                   1016.893690   
std                NaN           NaN                  15565.692133   
min                NaN           NaN                      0.000000   
25%                NaN           NaN                    149.000000   
50%                NaN           NaN                    190.000000   
75%                NaN           NaN                    236.000000   
max                NaN           NaN                 818738.000000   

        Songwriters_Listed_(1=Y_0=N) Song_Length Length_Compared_to_Median  
count                  100000.000000      100000                    100000  
unique                           NaN           8                         3  
top                              NaN     2-4 min                     Lower  
freq                             NaN       64179                     49891  
mean                        0.541610         NaN                       NaN  
std                         0.498268         NaN                       NaN  
min                         0.000000         NaN                       NaN  
25%                         0.000000         NaN                       NaN  
50%                         1.000000         NaN                       NaN  
75%                         1.000000         NaN                       NaN  
max                         1.000000         NaN                       NaN

## Visualizations & Insights

In [ ]:
sns.swarmplot(x=spotify.Song_Length, y=spotify.Streams)
plt.title('Spotify: Streams Per Song Length');

In [ ]:
plt.bar(spotify.Song_Length, spotify.Streams)
plt.show()

In [ ]:
plt.hist(mlc.Song_Length)
plt.xlabel('Song Length')
plt.ylabel('Quantity')
plt.title('Distribution of Song Length, Spotify')
plt.show()